tybalt/models.py

```diff
@@ -157,17 +193,68 @@ def compress(self, df):
                                   index=df.index)
         return encoded_df
 
-    def get_decoder_weights(self):
-        # build a generator that can sample from the learned distribution
-        # can generate from any sampled z vector
-        weights = []
-        for layer in self.decoder.layers:
-            weights.append(layer.get_weights())
-        return(weights)
 
-    def predict(self, df):
-        return self.decoder.predict(np.array(df))
+class Adage(BaseModel):
+    """
+    Training and evaluation of an ADAGE model
 
-    def save_models(self, encoder_file, decoder_file):
-        self.encoder.save(encoder_file)
-        self.decoder.save(decoder_file)
+    Usage: from tybalt.models import Adage
+    """
+    def __init__(self, original_dim, latent_dim, noise=0.05, batch_size=50,
+                 epochs=100, sparsity=0, learning_rate=1.1):
+        BaseModel.__init__(self)
+        self.original_dim = original_dim
+        self.latent_dim = latent_dim
+        self.noise = noise
+        self.batch_size = batch_size
+        self.epochs = epochs
+        self.sparsity = sparsity
+        self.learning_rate = learning_rate
+
+    def initialize_model(self):
+        """
+        Helper function to run that builds and compiles Keras layers
+        """
+        self.build_graph()
+        self.connect_layers()
+        self.compile_adage()
+
+    def build_graph(self):
+        # Build the Keras graph for an ADAGE model
+        self.input_rnaseq = Input(shape=(self.original_dim, ))
+        drop = Dropout(self.noise)(self.input_rnaseq)
+        self.encoded = Dense(self.latent_dim,
+                             activity_regularizer=l1(self.sparsity))(drop)
+        activation = Activation('relu')(self.encoded)
+        decoded_rnaseq = Dense(self.original_dim,
+                               activation='sigmoid')(activation)
+
+        self.full_model = Model(self.input_rnaseq, decoded_rnaseq)
+
+    def connect_layers(self):
+        # Separate out the encoder and decoder model
+        self.encoder = Model(self.input_rnaseq, self.encoded)
+
+        encoded_input = Input(shape=(self.latent_dim, ))
+        decoder_layer = self.full_model.layers[-1]
+        self.decoder = Model(encoded_input, decoder_layer(encoded_input))
+
+    def compile_adage(self):
```

Comment:
Do you want to give yourself the option to use different loss functions here? I see that you used MSE in [`adage_training.ipynb`](https://github.com/greenelab/tybalt/blob/0cebc8e71d25e5b568ad4329c3c2082f9f61d638/adage_training.ipynb)

Reply:
great point. Updated in 5178302